src/kernel/include/ulk.h

```diff
@@ -0,0 +1,48 @@
+#ifndef ULK_H
+#define ULK_H 
+
+#include "component.h"
+#include "cap_ops.h"
+#include "pgtbl.h"
+
+#define ULK_PGTBL_FLAG (1ul << 59)
+
+struct cap_ulk {
+    struct cap_header     h;
+    struct liveness_data  liveness;
+    vaddr_t               kern_addr;
+} __attribute__((packed));
+
+static int
+ulk_activate(struct captbl *t, capid_t ctcap, capid_t ulkcapid, livenessid_t lid, vaddr_t kaddr, capid_t ptcid, vaddr_t uaddr)
+{
+    struct cap_ulk   *ulkcap;
+    struct cap_pgtbl *ptcap;
+    paddr_t           pframe = chal_va2pa((void *)kaddr);
+    int               ret;
+
+    ptcap = (struct cap_pgtbl *)captbl_lkup(t, ptcid);
+    if (!ptcap || ptcap->h.type != CAP_PGTBL) return -EINVAL;
+    if (pgtbl_mapping_add(ptcap->pgtbl, uaddr, pframe, PGTBL_USER_DEF | ULK_PGTBL_FLAG, 12)) return -EINVAL;
+
+    ulkcap = (struct cap_ulk *)__cap_capactivate_pre(t, ctcap, ulkcapid, CAP_ULK, &ret);
+    if (!ulkcap) return -EINVAL;
+
+    ltbl_get(lid, &ulkcap->liveness);
+    ulkcap->kern_addr = kaddr;
```

Comment:
Should also store the uaddr (for the eventual removal).

Reply:
There's a challenging design issue here. We'll need to support *multiple* pages as there are more threads than their ulis can fit into a single page. So we'll need to change the thread creation path as well, and store the kernel address in the thread structure.

The code you have here is likely perfectly fine for that. But it is yet another capability (to this page) to pass into the thread creation. Worth discussing to make sure that we have a decent plan.

Note, you might have already solved this, and I simply haven't gotten to it in your review. Ignore if that's the case ;-)